Site Editor: Add navigation panel with placeholder content

david-szabo97 opened the pull request:
Related: https://github.com/WordPress/gutenberg/issues/25260 https://github.com/WordPress/gutenberg/issues/25261 

## Description
Putting Navigation component in action! Great way to see how it cooperates with the site editor. Also acts an example of how we could implement the preview popup.

This is a minimal implementation to see the Navigation component in the Site Editor.

## Needs fix
- [x] Menu title text color is overridden by other CSS (https://github.com/WordPress/gutenberg/pull/25507)
- [x] Group title text color is overridden by other CSS (https://github.com/WordPress/gutenberg/pull/25507)

## How has this been tested?
* `yarn wp-env start`
* Enable FSE
* Open Site Editor
* Click site logo to toggle the navigation sidebar

## Screenshots
Closed:
![image](https://user-images.githubusercontent.com/2256104/93774870-04edc380-fc22-11ea-9d15-324b0e54ab56.png)

Open:
![image](https://user-images.githubusercontent.com/2256104/93877113-345f0780-fcd8-11ea-8c44-c9e446e9b29e.png)

Preview:
![image](https://user-images.githubusercontent.com/2256104/93877141-404ac980-fcd8-11ea-8f86-ccbcbdd6282b.png)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

david-szabo97 committed:
Add navigation sidebar

vindl commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -89,7 +91,10 @@ export default function Header( {
 		<div className="edit-site-header">
 			<div className="edit-site-header_start">
 				<MainDashboardButton.Slot>
-					<FullscreenModeClose />
```
This button is exported from the package (experimental for now) and provides a way for 3rd plugins to customize some of its parts while retaining functionality ([see docs here](https://github.com/WordPress/gutenberg/blob/65258a077e4899ee4933c79077108f56f8b5ce6d/docs/designers-developers/developers/slotfills/main-dashboard-button.md)). Given that it will no longer represent a close action, I think that we should rename the existing `FullscreenModeClose` to `NavigationButton ` or `DashboardButton` and place our code changes there.

Copons replied:
We should also figure out how to handle the FSE Navigation when not in fullscreen mode, where we wouldn't have the WP/Site Icon back button anymore.

vindl replied:
Good point @Copons! I don't consider it a blocker for this PR though, as it's something that we can figure out later.

vindl commented on packages/edit-site/src/components/editor/index.js:
```diff
@@ -39,15 +39,33 @@ import { SidebarComplementaryAreaFills } from '../sidebar';
 
 function Editor() {
-	const [ isInserterOpen, setIsInserterOpen ] = useState( false );
+	const [ isInserterOpen, _setIsInserterOpen ] = useState( false );
```
Provided that we want this behavior, it might make sense to collapse these states into one `leftSidebarContent` that can have on of three possible values for now: `null`, `navigation`, and `inserter`. With present setup adding another element that wants to toggle into this space (there were some plans for doing that with block navigation) will be cumbersome, since each element will have to be aware of others and the toggling logic will have to be updated everywhere.

It would also remove the need for current naming with leading underscore (`setIsInserterOpen` and `_setIsInserterOpen`), which I find hard to read.

david-szabo97 replied:
Done!

vindl commented on packages/edit-site/src/components/editor/index.js:
```diff
@@ -156,6 +174,32 @@ function Editor() {
 		[ page?.context ]
 	);
 
+	let leftSidebar = null;
```
Should we wrap this block into helper function or make it a separate component?

david-szabo97 replied:
Yep! I'd say go for a separate component. And if we are already there, we should extract inserter panel as a separate component as well. What do you think about this structure?

```
packages/edit-site/src/components/left-sidebar
packages/edit-site/src/components/left-sidebar/inserter-panel
packages/edit-site/src/components/left-sidebar/navigation-panel (note: NavigationPanel instead of NavigationSidebar)
```

vindl replied:
Sounds fine. 👍

vindl reviewed (commented):
Nice work so far @david-szabo97! ✨

david-szabo97 commented:
#25507 is merged so I did a rebase. Updated the description with the new screenshots.

vindl commented:
@shaunandrews I have a question about the intended behavior here. Opening the inserter will close the navigation sidebar and vice versa in present version. Is that fine or should it be possible for both sidebars to be open at the same time?

Also, how should it interact with potential [Persistent block navigator panel](https://github.com/WordPress/gutenberg/issues/22113)?

david-szabo97 committed:
Replace is*Open states with a single state

david-szabo97 committed:
Extract left sidebar components

david-szabo97 committed:
Add back to dashboard item

david-szabo97 committed:
Change NavigationButton and FullscreenModalClose to DashboardButton

david-szabo97 commented:
- Extracted left sidebar components (`LeftSidebar`, `InserterPanel`, `NavigationPanel`)
- Added a temporary NavigationItem to the `NavigationPanel` which points to the dashboard
- Renamed `NavigationButton` to `DashboardButton` and removed `FullscreenModeClose`

david-szabo97 committed:
Fix export

vindl commented on packages/edit-site/src/components/left-sidebar/index.js:
```diff
@@ -0,0 +1,21 @@
+import NavigationPanel from './navigation-panel';
+
+const LeftSidebar = ( { content, setContent } ) => {
+	let leftSidebar = null;
```
I don't think we need `leftSidebar` var anymore or `else if`. We simplify this to:

```javascript
if ( content === 'navigation' ) {
    return <NavigationPanel />;
}

if ( content === 'inserter' ) {
    return <InserterPanel closeInserter={ () => setContent( null ) } />;
}

return null;

```

david-szabo97 replied:
Done! I was driven away by the original code 😓 😄

david-szabo97 committed:
Remove variable

vindl reviewed (approved):
Thanks @david-szabo97! This looks like a good starting point for our upcoming iteratons.

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:
```diff
@@ -0,0 +1,186 @@
+								title="Hover to show preview"
+								onMouseEnter={ () => setShowPreview( true ) }
+								onMouseLeave={ () => setShowPreview( false ) }
+							/>
```
How do you feel about removing all demo content except for "Back to dashboard" and "Hover to show preview"?
Everything else is just bloating the PR, and we can just look at the Storybook for usage examples.

david-szabo97 replied:
Removed them but I left a `NavigationGroup` just in case. Once this merged I'll put some real data in the navigation sidebar so this phase is kinda temporarily.

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:
```diff
@@ -0,0 +1,186 @@
+	const [ showPreview, setShowPreview ] = useState( false );
+
+	return (
+		<>
```
No need for this fragment here.

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:
```diff
@@ -0,0 +1,186 @@
+
+	return (
+		<>
+			<div style={ { position: 'relative', height: '100%' } }>
```
What about adding a `.edit-site-navigation-panel` class here, and move these styles to `styles.scss`?

Copons commented on packages/edit-site/src/components/header/dashboard-button/style.scss:
```diff
@@ -0,0 +1,42 @@
+.edit-site-dashboard-button_wrapper {
```
Please note that we should use a ["loose BEM" naming](https://github.com/WordPress/gutenberg/blob/f485f1f6a5556893b7774f2e9ef3320b0884c620/docs/contributors/coding-guidelines.md#css).
Modifier should be preceded by a double underscore.
E.g. `. edit-site-dashboard-button__wrapper`

(There are a few other occurrences of this in the PR, but I'll leave you the _fun_ of finding them 😛 )

david-szabo97 replied:
I was confused by `FullscreenModeClose` because that component is not following the BEM. 🤔 Thanks for letting me know, fixed it up!

Copons commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -89,7 +91,10 @@ export default function Header( {
+					<DashboardButton
+						isOpen={ isNavigationOpen }
+						onClick={ onToggleNavigation }
+					/>
```
I've got a few doubts about this change.
While it's true that this button doesn't close fullscreen mode anymore, it also doesn't navigate back to the dashboard (so the "Open dashboard" label in the component is misleading) — and it's not even technically _just_ a button 😄.

What it does now is to toggle the navigation sidebar, so I'd be more inclined in calling it `NavigationPanelToggle`.

---

I think we could also keep `FullscreenModeClose`.
While it would be left unused in this PR, we could update it to become (_be_?) the actual "dashboard button".
Consumers relying on it would not break, and we could use it inside the navigation toggle, or in the navigation itself (or wherever we see fit).

Either way, my opinion is that we should probably avoid removing existing components from an experimental PR whose purpose is to be a starting point, and not the finished product.

vindl replied:
I agree that that name is more appropriate. Also no objections to keeping the `FullscreenModeClose` if we are going to repurpose it and include it in navigation.

Copons replied:
My point is also that the navigation sidebar is still in flux, and I'm not super keen on removing an established component (albeit experimental).
We might end up not using it, or heavily modify it, but let's care about that when we'll be actually working on the back navigation. 🙂

david-szabo97 replied:
- Renamed `DashboardButton` to `NavigationToggle`
- Reverted `FullscreenModeClose` component, let's keep it for now and we will iterate upon it

Copons commented on packages/edit-site/src/components/left-sidebar/navigation-panel/index.js:
```diff
@@ -0,0 +1,186 @@
+
+const NavigationPanel = () => {
+	const [ activeItem, setActiveItem ] = useState( 'item-1' );
+	const [ activeMenu, setActiveMenu ] = useState( 'root' );
```
The `activeMenu` state is optional, only useful if we need to control the hierarchical navigation from outside.
It's not really needed here at this point.

Copons commented on packages/edit-site/src/components/editor/style.scss:
```diff
@@ -26,28 +26,7 @@
-.edit-site-editor__inserter-panel {
+.interface-interface-skeleton__left-sidebar .components-navigation {
+	width: 300px;
 	height: 100%;
```
This is probably not the right place for these styles.
What about:
- Remove everything from here.
- Create the `.edit-site-navigation-panel` class as suggested in [another comment](https://github.com/WordPress/gutenberg/pull/25506/files#r493329611).
- Add something like this in `packages/edit-site/src/components/left-sidebar/navigation-panel/style.scss`? 
```scss
.edit-site-navigation-panel {
	height: 100%;
	position: relative;
	width: 300px;

	.components-navigation {
		height: 100%;
	}
}
```

david-szabo97 committed:
Fix CSS naming

david-szabo97 committed:
Cleanup navigation content

david-szabo97 committed:
Cleanup CSS

david-szabo97 committed:
Fix width of navigation button

david-szabo97 committed:
Change hover color

david-szabo97 committed:
Remove fragment

david-szabo97 committed:
Rename DashboardButton to NavigationToggle

david-szabo97 committed:
Change label

david-szabo97 committed:
Fix export

david-szabo97 committed:
Revert FullscreenModeClose

david-szabo97 committed:
Add missing style

Copons reviewed (approved):
This LGTM!
Thank you @david-szabo97 for working on this! ✨ 

I'll open an issue for future discussion of the `FullscreenModeClose` destiny.

mattwiebe reviewed (approved):
This looks good to me. Left one piece of feedback that we can address in a renaming PR but nothing to block this.

afercia commented:
First observation: the `W` button was recently changed to accept the Site Icon set in the Customizer as a replacement for hte "W". In the block editor, a site icon looks like this:

<img width="902" alt="Screenshot 2020-09-24 at 18 02 27" src="https://user-images.githubusercontent.com/1682452/94170872-e04b5300-fe90-11ea-83bd-08039557e844.png">

In the Site Editor, it looks like this 🙂 : 

<img width="1075" alt="Screenshot 2020-09-24 at 18 02 05" src="https://user-images.githubusercontent.com/1682452/94171076-17216900-fe91-11ea-8016-cbd33218f24f.png">

where apparently there's some missing CSS.

Besides that, I'm not sure using the same UI control for different purposes on different screens is a good user experience. As a user, If I know the W button in the block editor brings me back to the Posts lists, I would expect the same button in the Site editor to do the same.

afercia commented:
Worth also noting the "sliding in sidebar" pattern is undergoing discussion by the accessibility team as it's a pattern that introduces various accessibility issues. It does relate to other new sidebars as well though so it's not specific to this PR. An audit is going to be made on the issue related to the main Inserter, that was already identified by the accessibility team as an accessibility regression in WordPress 5.5 compared to the implementation in 5.4. See #24975

Copons commented:
@afercia Thanks for chiming in!
The site icon size bug will be fixed in #25622.

The plan here was to merge the sidebar in the Site Editor in order to work on it in a real scenario, but still "protected" by the fact that it's an experimental environment.

We have been rushing this PR in because we (as in: people working on the Site Editor) are trying to improve the whole UX, which is arguably not very good right now.
One of the steps is to move the template switcher out of the header toolbar and into this sidebar. This will unblock other unrelated work that needs more room in the header toolbar, while keeping the template switching functionality available.

Thank you for linking that accessibility issue.
Admittedly, I haven't had time to look into it (I'm not an a11y expert, so it will take some time 🙂), but I promise we'll get to it.
